src/utils/common.ts

```diff
@@ -268,28 +268,32 @@ export async function estimateTotalGasRequiredByUnsignedTransaction(
   const { chainId } = await provider.getNetwork();
   const voidSigner = new VoidSigner(senderAddress, provider);
 
+  // Estimate the Gas units required to submit this transaction.
+  let gasCost = await voidSigner.estimateGas(unsignedTx);
+  let gasTokenCost: BigNumber;
+
   // OP stack is a special case; gas cost is computed by the SDK, without having to query price.
   if (chainIsOPStack(chainId)) {
     assert(isOptimismL2Provider(provider), `Unexpected provider for chain ID ${chainId}.`);
     assert(gasPrice === undefined, `Gas price (${gasPrice}) supplied for Optimism gas estimation (unused).`);
     const populatedTransaction = await voidSigner.populateTransaction(unsignedTx);
-    return (await provider.estimateTotalGasCost(populatedTransaction))
-      .mul(gasTotalMultiplier)
-      .div(toBNWei(1))
-      .toString();
-  }
-
-  if (!gasPrice) {
-    const gasPriceEstimate: GasPriceEstimate = await getGasPriceEstimate(provider);
-    gasPrice = gasPriceEstimate.maxFeePerGas;
+    gasTokenCost = await provider.estimateTotalGasCost(populatedTransaction);
+  } else {
+    if (!gasPrice) {
+      const gasPriceEstimate = await getGasPriceEstimate(provider);
+      gasPrice = gasPriceEstimate.maxFeePerGas;
+    }
+    gasTokenCost = gasCost.mul(gasPrice);
   }
 
-  // Estimate the Gas units required to submit this transaction
-  const estimatedGasUnits = await voidSigner.estimateGas(unsignedTx);
+  // Scale the results by the computed multiplier.
+  gasCost = gasCost.mul(gasTotalMultiplier).div(fixedPointAdjustment);
+  gasTokenCost = gasTokenCost.mul(gasTotalMultiplier).div(fixedPointAdjustment);
 
-  // Find the total gas cost by taking the product of the gas price & the
-  // estimated number of gas units needed.
-  return BigNumber.from(gasPrice).mul(gasTotalMultiplier).mul(estimatedGasUnits).div(toBNWei(1)).toString();
+  return {
+    gasCost, // Estimate in units of gas
+    gasTokenCost, // Estimate in units of native gas token
```

Comment:
So what's the difference between `gasCost` and `gasTokenCost`? `gasTokenCost` = `gasCost` * `gasUnitPriceUsd`?